pb/sigchain.go

```diff
@@ -361,22 +230,31 @@ func (sc *SigChain) finalSigElem() (*SigChainElem, error) {
 }
 
 func (sc *SigChain) IsFinal() bool {
-	if len(sc.Elems) < 2 || !common.IsEqualBytes(sc.DestPubkey, sc.Elems[len(sc.Elems)-2].NextPubkey) {
+	if sc.Length() < 3 {
 		return false
 	}
-	return true
+	pk := sc.Elems[len(sc.Elems)-2].NextPubkey
+	if pk == nil || bytes.Equal(pk, sc.DestPubkey) {
```

Comment:
`if pk == nil` is unnecessary because bytes.Equal() had handle it already.
Just `return bytes.Equal(pk, sc.DestPubkey)` directly is enough.

Reply:
Actually it's different. What you meant is `pk != nil && ...` while what we have is `pk == nil || ...`.

We allow `pk == nil` here because in order to save space, dest pk is not saved in relay packet, so the last node may not know the receiver client's pk. So we allow the last node's next pk to be nil to indicate that the next hop is the receiver client.

Reply:
Oh. Your expectation is `return true` when `pk == nil`.
How about `return pk == nil || bytes.Equal(pk, sc.DestPubkey)`?
`if true then return true` looks verbose.

Reply:
Agree, just changed it